File: 142.环形链表-ii.py
#
# @lc app=leetcode.cn id=142 lang=python3
#
# [142] 环形链表 II
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def detectCycle(self, head: ListNode) -> ListNode:
        if head == None or head.next == None:
            return None

        fast, slow = head, head
        while True:
            if not (fast and fast.next):
                return None
            slow, fast = slow.next, fast.next.next
            if slow == fast:
                break
        
        fast = head
        while fast != slow:
            slow, fast = slow.next, fast.next

        return fast
        
# @lc code=end

File: 142.环形链表-ii.cpp
/*
 * @lc app=leetcode.cn id=142 lang=cpp
 *
 * [142] 环形链表 II
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        if (head == nullptr || head->next == nullptr)
        {
            return nullptr;
        }

        ListNode *fast = head, *slow = head;
        while (true)
        {
            if (!(fast && fast->next)) {
                return nullptr;
            }
            slow = slow->next; 
            fast = fast->next->next;
            if (slow == fast) {
                break;
            }
        }

        fast = head;
        while (fast != slow)
        {
            fast = fast->next;
            slow = slow->next;
        }
        return fast;
        
    }
};
// @lc code=end

